### PR TITLE
Adding DualSense basic support

### PR DESCRIPTION
**Pull request by jimon**
### Description

Adding basic DualSense support.
Tested:
- All buttons work on PC/macOS over USB/Bluetooth (note, mic button doesn't work over bluetooth unless switched into full mode)
- Motor rumble/lightbar works on PC/macOS over USB. Bluetooth support not implemented (and apparently was not implemented for DualShock 4)
- More testing pending.

### Changes made

Had to add `IEventPreProcessor` to hijack and change events in-place before they reach `onEvent` callback, this is because the device can report at least 3 different report types which are not binary compatible with each other, and it can switch between them in runtime. The only architectural dead-end here is that we can only shrink events in-place. So when it will be time to add accelerometer/gyro support this will bite us.

### CODE REVIEW

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1404) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/InputSystem?pullRequest=1404) before we can accept your contribution.<br/><hr/>**Dmytro Ivanov** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/InputSystem?pullRequest=1404) it.</sub>

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs [hunk 1]
```diff
@@ -3148,6 +3152,15 @@ private unsafe void OnUpdate(InputUpdateType updateType, ref InputEventBuffer ev
                     }
                 }
 
+                // Give the device a chance to do something with data before we propagate it to event listeners.
```
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/InputManager.cs]: Wonder could we clean this bit up a little, like


```
  if (!skipEvent && device.hasEventPreProcessor && !((IEventPreProcessor)device).PreProcessEvent(currentEventReadPtr))
  {
    // Skip event if PreProcessEvent considers it to be irrelevant.
    m_InputEventStream.Advance(false);
    continue;
  }
```
I'm losing track of all the skipping and the leaving and whatnot. Seems clearer to do it in place.

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs [hunk 2]
```diff
@@ -3148,6 +3152,15 @@ private unsafe void OnUpdate(InputUpdateType updateType, ref InputEventBuffer ev
...
+                if (!skipEvent && device.hasEventPreProcessor)
```
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/InputManager.cs]: Should we add some tests for preprocess logic?

File: Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs [hunk 3]
```diff
@@ -203,6 +333,324 @@ public static DualShockHIDOutputReport Create()
...
+            if (eventPtr.type != StateEvent.Type)
+                return eventPtr.type != DeltaStateEvent.Type; // only skip delta state events
+
+            var stateEvent = StateEvent.FromUnchecked(eventPtr);
+            var size = stateEvent->stateSizeInBytes;
+            if (stateEvent->stateFormat != DualSenseHIDGenericInputReport.Format || size < sizeof(DualSenseHIDInputReport))
+                return false; // skip unrecognized state events otherwise they will corrupt control states
+
+            var genericReport = (DualSenseHIDGenericInputReport*)stateEvent->state;
+            if (genericReport->reportId == 0x01)
```
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs]: Could we put the magic numbers behind some well named constants here? 0x01, 0x31, and the two state event byte sizes 10 and 78

File: Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs [hunk 4]
```diff
@@ -203,6 +333,324 @@ public static DualShockHIDOutputReport Create()
...
+            {
+                if (stateEvent->stateSizeInBytes == 10 || stateEvent->stateSizeInBytes == 78)
+                {
+                    // minimal report
+                    var data = ((DualSenseHIDMinimalInputReport*)stateEvent->state)->ToHIDInputReport();
```
Comment (andrew-oc) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs:536]: Wondering if we need some validation here that what was returned from ToHIDInputReport isn't larger than the original size. Are there any guarantees that someone won't modify DualSenseHIDInputReport without realising this is a consequence?

**Review by andrew-oc — COMMENTED** [review]
Haven't tested on a device since I don't own a DualSense, but this looks mostly good to me. Just a few things to possibly tweak.

File: Packages/com.unity.inputsystem/InputSystem/Devices/IEventPreProcessor.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    /// <summary>
+    /// Gives an opportunity for device to modify event data in-place before it gets propagated to the rest of the system.
+    /// Beware that currently events can only shrink or stay the same size.
+    /// </summary>
+    /// <remarks>
+    /// If device also implements <see cref="IEventMerger"/> it will run first, because we don't process events ahead-of-time.
+    /// </remarks>
+    internal interface IEventPreProcessor
+    {
+        // return false to skip the event
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Devices/IEventPreProcessor.cs]: Minor: Why not move this return convention into summary comment if its part of the interface i.e. <returns>

File: Packages/com.unity.inputsystem/InputSystem/Editor/Debugger/InputDeviceDebuggerWindow.cs [hunk 6]
```diff
@@ -329,21 +329,23 @@ private void InitializeWith(InputDevice device)
         private void UpdateDeviceFlags()
         {
             var flags = new List<string>();
-            if ((m_Device.m_DeviceFlags & InputDevice.DeviceFlags.Native) == InputDevice.DeviceFlags.Native)
+            if (m_Device.native)
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/Debugger/InputDeviceDebuggerWindow.cs:332]: A lot easier to read and more difficult to do it wrong, awesome.

File: Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs [hunk 7]
```diff
@@ -203,6 +333,324 @@ public static DualShockHIDOutputReport Create()
...
+        {
+            var lf = lowFrequency.HasValue ? lowFrequency.Value : 0;
+            var hf = highFrequency.HasValue ? highFrequency.Value : 0;
+            var c = color.HasValue ? color.Value : Color.black;
+            
+            // DualSense differs a bit from DualShock 4 because all effects need to be set at a same time,
+            // otherwise setting a color would disable motor rumble.
+            var payload = new DualSenseHIDOutputReportPayload
+            {
+                enableFlags1 = 0x1 | 0x2,
```
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs]: Are these known flags (enum makes sense?) or just "magic numbers"?

File: Packages/com.unity.inputsystem/CHANGELOG.md [hunk 8]
```diff
@@ -8,6 +8,12 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 Due to package verification, the latest version below is the unpublished version and the date is meaningless.
 however, it has to be formatted properly to pass verification tests.
 
+## [Unreleased]
+
+### Added
+
+- Added support for DualSense gamepads on desktops (`DualSenseGamepadHID` layout).
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/CHANGELOG.md]: Might be useful to include PS5 in the name here. Just for extra clarity. Maybe

```
- Added support for PS5 DualSense controllers on Mac and Windows.
```

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs (hunk 2 above)
Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/InputSystem/InputManager.cs]: +100

File: Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepad.cs [hunk 9]
```diff
@@ -9,9 +9,9 @@
 namespace UnityEngine.InputSystem.DualShock
 {
     /// <summary>
-    /// A Sony DualShock controller.
+    /// A Sony DualShock/DualSense controller.
     /// </summary>
-    [InputControlLayout(displayName = "PS4 Controller")]
+    [InputControlLayout(displayName = "Playstation Controller")]
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepad.cs]: Hah, thank you! Stumbled over the same thing today and thought "WTF is there no PlayStation controller!"

Ah, should read `PlayStation`, not `Playstation`.

**Review by Rene-Damm — CHANGES_REQUESTED** [review]
Tests plz.

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs (hunk 1 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/InputManager.cs]: fixed

File: Packages/com.unity.inputsystem/CHANGELOG.md (hunk 8 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/CHANGELOG.md]: fixed

File: Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepad.cs (hunk 9 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepad.cs]: fixed

File: Packages/com.unity.inputsystem/InputSystem/Devices/IEventPreProcessor.cs (hunk 5 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Devices/IEventPreProcessor.cs]: fixed

File: Packages/com.unity.inputsystem/InputSystem/InputManager.cs (hunk 2 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/InputManager.cs]: fixed

File: Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs (hunk 4 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs:536]: I've added a check and a test in IEventPreProcessor handler itself

File: Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs (hunk 3 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs]: fixed

File: Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs (hunk 7 above)
Comment (jimon) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockGamepadHID.cs]: added comments

**Review by ekcoh — APPROVED** [review]

**Comment by ekcoh** [comment]
Great you added tests, seems to be a failing one in the CI though

**Review by andrew-oc — APPROVED** [review]

**Comment by jimon** [comment]
iOS CI just broke today, but `059a5fe` passed it, and `1df5163` only contains doc changes.
I'll revisit docs regarding UWP support in next PR

**Comment by Duf1220** [comment]
Hi. It will be available for InputDeviceTester? Thank You.

**Comment by jimon** [comment]
@Duf1220 should already be working in InputDeviceTester

**Pull request merged**